tests/federation_test.py

```diff
@@ -0,0 +1,253 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import datetime
+import logging
+import socket
+
+import absl
+from absl import flags
+from absl.testing import absltest
+from google.protobuf import message
+from grpc_channelz.v1 import channelz_pb2
+
+import framework
+import framework.helpers.docker
+import framework.helpers.logs
+import framework.helpers.retryers
+import framework.helpers.xds_resources
+import framework.xds_flags
+import framework.xds_k8s_testcase
+
+logger = logging.getLogger(__name__)
+
+_CONTROL_PLANE_IMAGE = flags.DEFINE_string(
+    "control_plane_image",
+    "us-docker.pkg.dev/grpc-testing/psm-interop/test-control-plane:latest",
+    "Control plane (xDS config) server image",
+)
+_HOST_NAME = flags.DEFINE_string(
+    "host_name",
+    "host.docker.internal",
+    "Host name all the services are bound on",
+)
+_NODE_ID = flags.DEFINE_string("node", "test-id", "Node ID")
+_STATUS_TIMEOUT_MS = flags.DEFINE_integer(
+    "status_timeout_ms",
+    15000,
+    "Duration (in ms) that the test will wait for xDS channel to change the status",
+)
+_STATUS_POLL_INTERVAL_MS = flags.DEFINE_integer(
+    "status_poll_interval_ms", 300, "Channel status poll interval (in ms)"
+)
+_STATS_REQUEST_TIMEOUT_S = flags.DEFINE_integer(
+    "stats_request_timeout_s",
+    300,
+    "Number of seconds the client will wait for the requested number of RPCs",
+)
+_LISTENER = "listener_0"
+
+absl.flags.adopt_module_key_flags(framework.xds_k8s_testcase)
+
+
+def get_free_port() -> int:
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        sock.bind(("localhost", 0))
+        return sock.getsockname()[1]
+
+
+class FederationTest(absltest.TestCase):
+    bootstrap: framework.helpers.docker.Bootstrap = None
+    dockerInternalIp: str
+    authority1_port: int
+    authority2_port: int
+
+    @staticmethod
+    def setUpClass():
+        # Use the host IP for when we need to use IP address and not the host
+        # name, such as EDS resources
+        FederationTest.dockerInternalIp = socket.gethostbyname(
+            socket.gethostname()
+        )
+        FederationTest.authority1_port = get_free_port()
+        FederationTest.authority2_port = get_free_port()
+        authorities = {
+            "authority1": f"{_HOST_NAME.value}:{FederationTest.authority1_port}",
+            "authority2": f"{_HOST_NAME.value}:{FederationTest.authority2_port}",
+        }
+        server_template = "xdstp://authority2/envoy.config.listener.v3.Listener/grpc/server/%s"
+        FederationTest.bootstrap = framework.helpers.docker.Bootstrap(
+            framework.helpers.logs.log_dir_mkdir("bootstrap"),
+            # Use an invalid domain for the default server, to validate that no
+            # endpoint tries to use it
+            servers=["invalid.:8080"],
+            authorities=authorities,
+            server_template=server_template,
+        )
```

Comment:
Consider moving to `setUp()` instead - it runs before each test. If we ever add another `_test` method, we'll run in the port conflicts